File: Creational/Prototype/PrototypePatternRunner.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Pattern.Creational.Prototype
{
	internal static class PrototypePatternRunner
	{
		internal static void Run()
		{
			Console.WriteLine("Prototype Pattern");
			var original = new ConcretePrototype("Augusta Ada Lovelace", 1);
			Console.WriteLine("Original Prototype:");
			Console.WriteLine(original.ToString());
			
			var clone = (ConcretePrototype) original.Clone();
			Console.WriteLine("Cloned Prototype:");
			Console.WriteLine(clone.ToString());
			
			clone.Name = "Max Schmeling";
			clone.Number = 2;
			Console.WriteLine("Cloned Modified Prototype:");
			Console.WriteLine(clone.ToString());
		}
	}
}

File: VB/Creational/Prototype/PrototypePatternRunner.vb
'
' 1.3 Prototype
'
' Von einem zu kopierenden Objekt kann nicht immer auf alle
' Eigenschaften zugegriffen werden, zb weil sie private sind. Daher Kopiervorgang im Objekt selbst durchführen.
' Neben dem normalen Konstruktur mit Method Overloading einen weiteren Konstruktur erstellen
' der das Objekt selbst annimmt und in das neue Objekt die Werte des übergebenden Objekts schreibt
' Für häufig kopierte Objekt kann eine PrototypeRegistry angelegt werden, die Objekte zwischenspeichert und
' auf Anfrage einen Klon erstellt und ausgibt.
' 
Public Class PrototypePatternRunner
	Public Sub New()
	End Sub

	Public Function Run()
		Console.WriteLine("Prototype Pattern..")

		Dim original As ConcretePrototype = New ConcretePrototype("Ada Lovelace", 1)
		Console.WriteLine("original:")
		Console.WriteLine(original.ToString())

		Dim clone As ConcretePrototype = original.Clone()
		Console.WriteLine("cloned:")
		Console.WriteLine(clone.ToString())

		clone.Name = "Max Schmeling"
		clone.Number = 2
		Console.WriteLine("cloned modified:")
		Console.WriteLine(clone.ToString())


	End Function
End Class
